ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,70 @@
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some other possibilities might evolve later. Above is just an example used now by some Performance teams.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in StoneSoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
+
+### Glossary
+
+- SUT: Subject Under Test
+- Horreum: a results storage service: <https://github.com/Hyperfoil/Horreum>
+- CPT: Continuous Performance Testing
+
+## Decision
+
+Let's use this architecture with a single Horreum instance that would be managed by Red Hat and used by all customers.
+
+![Architecture diagram with Horreum](assets/0018-apps-continuous-perf-testing.svg "Architecture diagram with Horreum")
+
+1. Performance test runs in the Tekton pipeline and generates JSON with test parameters and results.
+2. Pipeline gathers configured monitoring metrics and add them to the JSON.
+3. Pipeline uploads the JSON with all the data to Horreum.
+4. Horreum performs result analysis, looking for changes in configured metrics.
+4. Pipeline gets PASS/FAIL decision from Horreum back to the pipeline, so pipeline can return proper result.
+
+## Consequences
+
+Pros:
+
+- We can provide some best practices to customers wanting to do some basic performance regression testing.
+- Horreum is end-to-end solution that already exists, is already used by multiple teams and has documentation for the whole process.
+- StoneSoup UI can display all the trends because historical data will be available in Horreum.
+- Red Hat team that develops Horreum is willing to help with this effort.
+- Integration scripts that would be running in the pipeline (to gather monitoring data, to upload results to Horreum, to get PASS/FAIL decision from Horreum...) mostly exists.
+
+Cons:
+
+- Somebody has to manage the Horreum instance. Required development time on it. Service instance needs ownership.
```

Comment:
We need to resolve this (somehow) before moving forwards much more. We can't introduce a new subsystem without clear ownership of updating it, accepting updates to it, uptime, etc.

Reply:
@ralphbean   
Are you asking about production ownership or Development efforts? 
I thought AppSRE will take ownership of maintaining these services.

Reply:
AppSRE won't take ownership of maintaining services. All services will be maintained with a hybrid SRE model with shared responsibility between SRE and the workload teams.